### DISCUSSION TITLE
Spatial Derivative

### DISCUSSION THREAD

**Discussion by Josh-Web-700** [idaholab/moose#27792] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Hello,

I am attempting to implement a kernel in MOOSE that contains the directional derivative (d/dx) of a material property. Is there a way to take the gradient or directional derivative of a material property inside the kernel and/or the materials system? If not, what is the best way to implement this?

It seems like this would be straightforward to do in an AuxKernel, but the material property is dependent on the gradient of the kernel variable (_grad_u) and should be compatible with automatic differentiation (AD) to be computed inside the kernel.

Another solution would be to use adCoupledGradient() to retrieve the gradient of this material property, and while I am new to MOOSE, from my understanding this applies only to coupled variables and not to material properties.

Thank you! Let me know if you need any more information or clarification.

**Comment by GiudGiud** [comment, marked as answer]
Hello

No the easiest and cleanest way to do this is to define the spatial derivative (gradient) of the material property in either the same Material object or in a new one. 

The other clean way of doing this is reworking the material to use the derivativeMaterialInterface to compute these derivatives automatically

**Reply by Josh-Web-700** [reply]
Thank you! That reinforces the suggestions you made in Discussion #20628, and that is very helpful.
